File: lib/app/data/models/plan_item.dart
import 'package:freezed_annotation/freezed_annotation.dart';
import 'item_unit.dart';

part 'plan_item.freezed.dart';
part 'plan_item.g.dart';

enum PlanStatus {
  pending,
  completed;

  String get displayName {
    switch (this) {
      case PlanStatus.pending:
        return 'Pending';
      case PlanStatus.completed:
        return 'Completed';
    }
  }
}

@freezed
class PlanItem with _$PlanItem {
  const factory PlanItem({
    required String id,
    required String title,
    required DateTime date,
    String? time,
    @Default('') String category,
    @Default(PlanStatus.pending) PlanStatus status,
    String? linkedQuestId,
    String? notes,
    // Quantity and unit tracking (for tasks like "Walk 10 minutes", "Do 3 sets")
    @Default(0) int quantity, // Quantity (e.g., 10 for "10 minutes", 3 for "3 sets")
    @Default(ItemUnit()) ItemUnit unit, // Unit type (e.g., "minutes", "sets", "reps")
  }) = _PlanItem;

  factory PlanItem.fromJson(Map<String, dynamic> json) =>
      _$PlanItemFromJson(json);
}

File: backend/app/models/plan_item.py
"""PlanItem model matching Flutter's plan_item.dart."""

from datetime import datetime
from enum import Enum
from pydantic import BaseModel, Field

from .common import ItemUnit


class PlanStatus(str, Enum):
    """Plan item status."""
    pending = "pending"
    completed = "completed"


class PlanItemBase(BaseModel):
    """Base plan item fields."""
    title: str = Field(..., min_length=1, max_length=200)
    date: datetime
    time: str | None = None
    category: str = ""
    status: PlanStatus = PlanStatus.pending
    linked_quest_id: str | None = None
    notes: str | None = None
    quantity: int = 0
    unit: ItemUnit = Field(default_factory=ItemUnit)


class PlanItemCreate(PlanItemBase):
    """Plan item creation request."""
    pass


class PlanItemUpdate(BaseModel):
    """Plan item update request."""
    title: str | None = Field(None, min_length=1, max_length=200)
    date: datetime | None = None
    time: str | None = None
    category: str | None = None
    status: PlanStatus | None = None
    linked_quest_id: str | None = None
    notes: str | None = None
    quantity: int | None = None
    unit: ItemUnit | None = None


class PlanItem(PlanItemBase):
    """Plan item response model."""
    id: str
    user_id: str
    created_at: datetime
    updated_at: datetime
    
    class Config:
        from_attributes = True
        use_enum_values = True

